File: Level1/실패율.py
N=5
stages=[2, 1, 2, 6, 2, 4, 3, 3]
def solution(N, stages):
    answer = {} #딕셔너리 이용
    clear_palyer=len(stages)
    for i in range(N):
        if clear_palyer>=1:
            cnt=stages.count(i+1) #i+1번째 스테이지 클리어한 넘 수 구함
            answer[i+1]=cnt/clear_palyer #실패율
            clear_palyer-=cnt #다음 스테이지는 i+1애들은 클리어 못했으므로
        else: #만약 i번째 스테이지를 클리어 못한 경우
            answer[i+1]=0
    answer = sorted(answer, key=lambda x: answer[x], reverse=True) #정렬 기준 value들의 오름차순
    return answer
print(solution(N,stages))

File: Level1/실패율.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <map>

using namespace std;

bool compare(pair<int, double> a, pair<int, double> b) {
    if (a.second == b.second)
        return a.first < b.first;
    else
        return a.second > b.second;
}

vector<int> solution(int N, vector<int> stages) {
    vector<int> answer;
    int num = stages.size();
    //각 플레이어의 스테이지 실패율
    map<int, int> dict;
    //실패율
    vector<pair<int, double>> fails;

    //스테이지 실패율을 체크
    //만약 KEY값이 N을 넘는 경우는 올 클리어를 했다는 뜻
    //실패율 구할 때 고려 X
    for (int i = 0; i < num; ++i)
        dict[stages[i]]++;

    for (int i = 1; i <= N; ++i) {
        //도달한 유저가 없는 경우는 실패율을 0으로 취급
        if (dict[i] == 0) {
            fails.push_back({ i, 0 });
            continue;
        }
        fails.push_back({ i, (double)dict[i] / num });
        num -= dict[i];
    }
    //실패율 기준으로 내림차순 정렬
    sort(fails.begin(), fails.end(), compare);

    //스테이지만 정답벡터에 삽입
    for (int i = 0; i < N; ++i)
        answer.push_back(fails[i].first);
    return answer;
}